homeassistant/components/conversation/default_agent.py

```diff
@@ -1181,17 +1247,62 @@ def _get_match_error_response(
 
     if reason == intent.MatchFailedReason.STATE:
         # Entity is not in correct state
-        assert match_error.constraints.states
-        state = next(iter(match_error.constraints.states))
-        if match_error.constraints.domains:
+        assert constraints.states
+        state = next(iter(constraints.states))
+        if constraints.domains:
             # Translate if domain is available
-            domain = next(iter(match_error.constraints.domains))
+            domain = next(iter(constraints.domains))
             state = translation.async_translate_state(
                 hass, state, domain, None, None, None
             )
 
         return ErrorKey.ENTITY_WRONG_STATE, {"state": state}
 
+    if reason == intent.MatchFailedReason.ASSISTANT:
+        # Not exposed
+        if constraints.name:
+            if constraints.area_name:
+                return ErrorKey.NO_ENTITY_IN_AREA_EXPOSED, {
+                    "entity": constraints.name,
+                    "area": constraints.area_name,
+                }
+            if constraints.floor_name:
+                return ErrorKey.NO_ENTITY_IN_FLOOR_EXPOSED, {
+                    "entity": constraints.name,
+                    "floor": constraints.floor_name,
+                }
+            return ErrorKey.NO_ENTITY_EXPOSED, {"entity": constraints.name}
+
+        if constraints.device_classes:
+            device_class = next(iter(constraints.device_classes))
+
+            if constraints.area_name:
+                return ErrorKey.NO_DEVICE_CLASS_IN_AREA_EXPOSED, {
+                    "device_class": device_class,
+                    "area": constraints.area_name,
+                }
+            if constraints.floor_name:
+                return ErrorKey.NO_DEVICE_CLASS_IN_FLOOR_EXPOSED, {
+                    "device_class": device_class,
+                    "floor": constraints.floor_name,
+                }
+            return ErrorKey.NO_DEVICE_CLASS_EXPOSED, {"device_class": device_class}
+
+        if constraints.domains:
+            domain = next(iter(constraints.domains))
+
+            if constraints.area_name:
+                return ErrorKey.NO_DOMAIN_IN_AREA_EXPOSED, {
+                    "domain": domain,
+                    "area": constraints.area_name,
+                }
+            if constraints.floor_name:
+                return ErrorKey.NO_DOMAIN_IN_FLOOR_EXPOSED, {
+                    "domain": domain,
+                    "floor": constraints.floor_name,
+                }
+            return ErrorKey.NO_DOMAIN_EXPOSED, {"domain": domain}
+
```

Comment:
I'm not sure I understand how this works.

Just because you have `constraint.name` does not mean that this is where the matching failed. It could have failed later. Or earlier.

What is the selection logic behind this code?

Reply:
If we're here, the match succeeded all of the other checks and failed the final exposure check (with the reworked logic). The goal is to give as much information to the user as we can so that they can expose the correct entity or entities.

If they provided a name (`constraint.name`), for example, the exposure check would only fail if there was an entity of that name but it wasn't exposed. Likewise with the `constraints.area_name` underneath: if an entity by that name wasn't in the area, we would have never reached the exposure check (failing instead with `MatchFailedReason.NAME`).

We have to be careful to put device class and domain last (in that order), because some intent sentences provide the domain when matched (e.g., https://github.com/home-assistant/intents/blob/db81682233a12e8d88296bd9c3929de78c7bc139/sentences/en/lock_HassTurnOn.yaml#L6). This will add to `constraints.domains`, but we only want to say there aren't any locks exposed if we don't have a `name` slot like here: https://github.com/home-assistant/intents/blob/db81682233a12e8d88296bd9c3929de78c7bc139/sentences/en/lock_HassTurnOn.yaml#L14

Reply:
With the reworked logic in `intent.py` this makes sense.